node/core/dispute-coordinator/src/import.rs

```diff
@@ -517,14 +519,17 @@ impl ImportResult {
 
 		let (mut votes, _) = new_state.into_old_state();
 
-		for (index, sig) in approval_votes.into_iter() {
+		for (index, (candidate_hashes, sig)) in approval_votes.into_iter() {
+			gum::debug!(
+				target: LOG_TARGET,
+				"Imported votes for {:?}", candidate_hashes
+			);
 			debug_assert!(
 				{
 					let pub_key = &env.session_info().validators.get(index).expect("indices are validated by approval-voting subsystem; qed");
-					let candidate_hash = votes.candidate_receipt.hash();
 					let session_index = env.session_index();
-					DisputeStatement::Valid(ValidDisputeStatementKind::ApprovalChecking)
-						.check_signature(pub_key, candidate_hash, session_index, &sig)
+					candidate_hashes.contains(&votes.candidate_receipt.hash()) && DisputeStatement::Valid(ValidDisputeStatementKind::ApprovalCheckingV2(candidate_hashes.clone()))
+						.check_signature(pub_key, *candidate_hashes.first().expect("Valid votes have at least one candidate; qed"), session_index, &sig)
```

Comment:
Shouldn't all hashes be provided as when signing ?

Reply:
It is provided in `ApprovalCheckingV2(candidate_hashes.clone())`.

But, yeah, I think you are right `check_signature` api is not obvious anymore, I'll see how I can fix it.